cpp/include/raft/core/mdarray.hpp

```diff
@@ -52,6 +53,54 @@ template <typename ElementType,
           typename AccessorPolicy = detail::stdex::default_accessor<ElementType>>
 using mdspan = detail::stdex::mdspan<ElementType, Extents, LayoutPolicy, AccessorPolicy>;
 
+using matrix_extent = detail::matrix_extent;
+
+using storage_order_type = detail::stdex::StorageOrderType;
+
+template <typename ElementType, storage_order_type order>
+using padded_layout = detail::stdex::layout_padded_general<
+  detail::stdex::padding<std::remove_cv_t<std::remove_reference_t<ElementType>>>::value,
+  order>;
+
+template <typename ElementType, storage_order_type order>
+using padded_matrix = mdspan<ElementType, matrix_extent, padded_layout<ElementType, order>>;
+
+// alignment_v<remove_cv_t<remove_reference_t<ElementType>>>
+struct alignment {
+  static constexpr size_t value = 128;
+};
+
+template <class ElementType, class Extents, storage_order_type order>
+using aligned_mdspan =
+  mdspan<ElementType,
+         Extents,
+         detail::stdex::layout_padded_general<
+           detail::stdex::padding<std::remove_cv_t<std::remove_reference_t<ElementType>>>::value,
+           order>,
+         detail::stdex::aligned_accessor<ElementType, alignment::value>>;
+
+template <class ElementType, class Extents, storage_order_type order>
+aligned_mdspan<ElementType, Extents, order> make_aligned_mdspan(ElementType* input_pointer,
+                                                                Extents e,
+                                                                storage_order_type /*order*/)
+{
+  using value_type = std::remove_cv_t<std::remove_reference_t<ElementType>>;
+  using pointer = typename detail::stdex::aligned_accessor<ElementType, alignment::value>::pointer;
+
+  // Any run-time check for alignment would happen here.
+  // Taking the alignment value as integral_constant permits interesting checks,
+  // like wrapping ElementType* in a proxy pointer type
+  // that encodes alignment as a non-type template parameter.
+  assert(input_pointer == alignTo(input_pointer, alignment::value));
+
+  pointer aligned_pointer = input_pointer;  // assert_aligned(input_pointer,
```

Comment:
[INFORMATIVE NOTE]

I've been thinking about how to express this idea.  The issue is that `std::assume_aligned` (C++20) or equivalent compiler built-ins (e.g., GCC's `__builtin_assume_aligned`) don't affect the type.  My current thinking is that `aligned_accessor::access` should do `return (std::assume_aligned(p))[i];`.  Please see my comment below on `aligned_accessor::access`.

Reply:
For my latest thoughts on `assume_aligned` etc., please see my `aligned_accessor` PR https://github.com/kokkos/mdspan/pull/176 .